pkg/virt-handler/backoff/backoff.go

```diff
@@ -0,0 +1,118 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package backoff
+
+import (
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type limitedBackoff struct {
+	backoff          wait.Backoff
+	initialBackoff   wait.Backoff
+	deadlineDuration time.Duration
+	deadline         time.Time
+}
+
+func newLimitedBackoff(deadline, initialBackoff, maxBackoff time.Duration, steps int, factor, jitter float64) *limitedBackoff {
+	limitedBackoff := limitedBackoff{
+		deadlineDuration: deadline,
+		initialBackoff: wait.Backoff{
+			Duration: initialBackoff,
+			Cap:      maxBackoff,
+			Steps:    steps,
+			Factor:   factor,
+			Jitter:   jitter,
+		},
+	}
+	limitedBackoff.init()
+	return &limitedBackoff
+}
+
+func (l *limitedBackoff) init() {
+	l.backoff = l.initialBackoff
+	l.deadline = time.Now().Add(l.deadlineDuration)
+}
+
+func (l *limitedBackoff) Reset() {
+	l.backoff = l.initialBackoff
+	l.deadline = time.Now().Add(l.deadlineDuration)
+}
+
+func (l *limitedBackoff) Step() time.Duration {
+	if time.Now().After(l.deadline) {
+		return time.Duration(0)
+	}
+	return l.backoff.Step()
+}
+
+type Stepper interface {
+	Step() time.Duration
+	Reset()
+}
+
+type backoffStepper struct {
+	backoff Stepper
+	start   time.Time
+	end     time.Time
+}
+
+func newBackoffStepper(backoff Stepper) *backoffStepper {
+	b := &backoffStepper{
+		backoff: backoff,
+	}
+	b.init()
+	return b
+}
+
+func (b *backoffStepper) init() {
+	b.backoff.Reset()
+	now := time.Now()
+	b.start = now
+	b.end = now
+}
+
+func (b *backoffStepper) Reset() {
+	b.init()
+}
+
+func (b *backoffStepper) NextBackoff() {
+	nextBackoff := b.backoff.Step()
+	if nextBackoff < 1 {
+		return
+	}
+
+	now := time.Now()
+	if time.Now().Before(b.end) {
+		return
+	}
+
+	b.start = now
+	b.end = b.start.Add(nextBackoff)
+}
+
+func (b *backoffStepper) BackoffEndTime() time.Time {
+	return b.end
+}
+
+func NewSriovHotplugScheduler() *backoffStepper {
```

Comment:
This does not seem to belong here. The `backoff` is not suppose to know about specific users of it.

Reply:
Done